### PR TITLE
Improve using cards within sections

### PR DESCRIPTION
**Pull request by warpech**
I made various fixes regarding using cards within sections and created a demo of that.

Changes are explained in the commit messages.

### CODE REVIEW

File: src/uniform.css [hunk 1]
```diff
@@ -26,7 +26,8 @@
   --warning: var(--uni-yellow);
   --disabled: var(--light-gray);
 
-  --uni-host-padding: 20px;
+  --uni-host-padding-horizontal: 40px;
```
Comment (alshakero) [inline src/uniform.css]: That's a huge padding to be standarized. That's 80px total. On mobile phones that's 25% of the screen. When I like bigger-than-usual paddings I usually use `1%`. That's `19.2px` on most desktop computers (1920 HD) and 3px on mobile phones. Looks great. In this case `2%` would be cool (38.4px on PC and 6px on phone).

File: index.html [hunk 2]
```diff
@@ -609,16 +620,14 @@ <h2>Sections</h2>
             Note, that those sections stick to the sides of padded host.
             <div data-shadow-host>
                 <template is="declarative-shadow-dom" mode="open">
+                    <style>@import 'uniform.css';</style>
                     <!--
                         To make a section stick to the top of a padded container,
                         it needs to be `:first-child`, put style tag below.
                     -->
```
Comment (alshakero) [inline index.html]: This comment is no longer valid I assume?

File: index.html [hunk 3]
```diff
@@ -628,24 +637,44 @@ <h2>Sections</h2>
...
-</section>
+        </section>
+        <section>
+            <h2>Section with card</h2>
+            You can nest cards within sections
+            <div data-shadow-host>
+                <template is="declarative-shadow-dom" mode="open">
+                    <style>@import 'uniform.css';</style>
+                    <div class="uni-section-primary">
+                        <div class="uni-card">Default card in a section (<code>inline-block</code>)</div>
```
Comment (alshakero) [inline index.html]: I would use "Default card in a section (`<code>display: inline-block</code>`)"

File: src/uniform.css [hunk 4]
```diff
@@ -345,50 +347,50 @@
...
-  margin-right: calc(-1 * var(--uni-host-padding));
+  margin-left: calc(-1 * var(--uni-host-padding-horizontal));
+  margin-right: calc(-1 * var(--uni-host-padding-horizontal));
 }
 :host > .uni-section-tertiary {
-  margin-left: calc(-1 * var(--uni-host-padding));
-  margin-right: calc(-1 * var(--uni-host-padding));
+  margin-left: calc(-1 * var(--uni-host-padding-horizontal));
+  margin-right: calc(-1 * var(--uni-host-padding-horizontal));
 }
```
Comment (alshakero) [inline src/uniform.css]: can't this all be
:host > .uni-section-tertiary,
:host > .uni-section-secondary ,
:host > .uni-section-primary {
  margin-left: calc(-1 * var(--uni-host-padding-horizontal));
  margin-right: calc(-1 * var(--uni-host-padding-horizontal));

File: index.html [hunk 5]
```diff
@@ -628,24 +637,44 @@ <h2>Sections</h2>
...
+                    </div>
+                    <div class="uni-section-secondary">
+                        <div class="uni-card" style="display: block">Full-width card in a section (<code>block</code>)</div>
```
Comment (alshakero) [inline index.html]: Same as above. Especially `block` is a very generic word.

File: index.html [hunk 6]
```diff
@@ -628,24 +637,44 @@ <h2>Sections</h2>
...
+                    </div>
+                    <div class="uni-section-primary" style="display: grid;">
+                        <div class="uni-card">Full-width card in a section (<code>grid</code>)</div>
```
Comment (alshakero) [inline index.html]: Same as above.

**Comment by alshakero** [comment]
I volunteered to review because I merged the last PR.

File: src/uniform.css (hunk 1 above)
Comment (warpech) [reply src/uniform.css]: That's a great comment for a future version, but these are the designs that we got from @andwah.

File: src/uniform.css (hunk 4 above)
Comment (warpech) [reply src/uniform.css]: I think yes, thanks! I have tested it in Firefox, Chrome and Safari.

**Comment by warpech** [comment]
Thanks @alshakero, all fixed.

File: src/uniform.css [hunk 7]
```diff
@@ -359,50 +367,48 @@ Pagination
...
 /* Let first section to stick to the top of the host */
-:host > .uni-section-primary:first-child,
-:host > .uni-section-secondary:first-child,
-:host > .uni-section-tertiary:first-child {
```
Comment (tomalec) [inline src/uniform.css]: Are you sure you want to limit it to first of type?

Comment (warpech) [reply src/uniform.css]: How about we limit section to be only usable with one element type? Such as `section` or `div`. It's easier to support such a narrow case.

**Comment by tomalec** [comment]
What is the rationale to have a separate section and demo of cards in sections? This is HTML so it's natural you can nest elements, like, you can put a button in a card in a, etc. I think we should not put the demos for all the combinations, as we could blur the picture.

If we **really** want to showcase how the section looks and adapts to different content, I'd put it in the same ekhm... section (`h2`) as "Sections", and add source code of it.

File: src/uniform.css (hunk 1 above)
Comment (alshakero) [reply src/uniform.css]: >Perhaps we can actually use values relative to the viewport size instead of pixels?

Yes that's what I meant.

Comment (alshakero) [reply src/uniform.css]: I meant using literally `padding: 2%`.

File: src/uniform.css (hunk 7 above)
Comment (alshakero) [reply src/uniform.css]: We can also do this.margined {
  background-color: red;
  height: 50px;
  margin: 5px;
  margin-top: 50px;
}

.margined:first-of-type:not(:first-child) {
   margin-top: 0px
}

**Comment by alshakero** [comment]
> all fixed apart from the 40/20px margins.

I think we unknowingly agreed to a valid solution? https://github.com/Starcounter/uniform.css/pull/33#discussion_r168639993.

**Comment by alshakero** [comment]
Even though I think all those duplicate selectors shouldn't be seperate. The most maintainence-prone part is the values not the selectors, having the values twice is a bit smelly.
```
.uni-alert-warning > ::slotted(*) {
  color: var(--dark);
  background-color: var(--warning);
  border-color: var(--warning);
}
.uni-alert-warning > * {
  color: var(--dark);
  background-color: var(--warning);
  border-color: var(--warning);
}
```

to
```
.uni-alert-warning > ::slotted(*),
.uni-alert-warning > * {
  color: var(--dark);
  background-color: var(--warning);
  border-color: var(--warning);
}
```

There is a lot of them.

**Comment by warpech** [comment]
> @alshakero [..] I think we unknowingly agreed to a valid solution? #33 (comment).

Moved to a new issue https://github.com/Starcounter/uniform.css/issues/36

> @tomalec [..] What is the rationale to have a separate section and demo of cards in sections? This is HTML so it's natural you can nest elements, like, you can put a button in a card in a, etc. I think we should not put the demos for all the combinations, as we could blur the picture.

1. Because @PatrykSzwer didn't know how to do it and asked for it
2. To stress out that they are intended to be used together
3. See Bootstrap: they have plenty of non-obvious combination examples: https://getbootstrap.com/docs/4.0/components/input-group/
4. Sometimes (usually?) a picture is blurry because there is *not enough* details, not because there is *too much* details.

> @tomalec [..] If we really want to showcase how the section looks and adapts to different content, I'd put it in the same ekhm... section (h2) as "Sections", and add source code of it.

Good idea!

> @alshakero [..] The readme note regarding duplication would remain valid though.

It is not an accident that these rules are duplicated. As the readme note explains, we MUST have this duplication, otherwise the polyfilled browsers will not be able to parse the stylesheet. I don't think there is anything we can do about it.

**Comment by tomalec** [comment]
WDYT about:
```css
/* Let first section to stick to the top of the host */
:host > .uni-section-primary:first-of-type,
:host > .uni-section-secondary:first-of-type,
:host > .uni-section-tertiary:first-of-type {
  margin-top: calc(-1 * var(--uni-host-padding-vertical));
}
:host > .uni-section-primary ~ .uni-section-primary,
:host > .uni-section-primary ~ .uni-section-secondary,
:host > .uni-section-secondary ~ .uni-section-primary,
/*...*/
:host > .uni-section-tertiary ~ .uni-section-primary {
  margin-top: 0;
}
```

**Comment by PatrykSzwer** [comment]
I would kindly ask for removing ` margin-bottom: 20px;` line from uni-alert class. With new design and uni-section classes It's bodering more than helping.

**Comment by alshakero** [comment]
> It is not an accident that these rules are duplicated. As the readme note explains, we MUST have this duplication, otherwise the polyfilled browsers will not be able to parse the stylesheet. I don't think there is anything we can do about it.

Yes I'm not arguing for removing duplicate selectors. Only duplicate values.

**Comment by tomalec** [comment]
> Yes I'm not arguing for removing duplicate selectors. Only duplicate values.

Could you give the example of before and after?

Polyfilled browsers require us to make duplicate rules, as once they fail to parse selector (for example due to the presence of unknown `::slotted`) they disregard entire rule (together with declaration block => values).


 See https://developer.mozilla.org/en-US/docs/Learn/CSS/Introduction_to_CSS/Syntax#A_touch_of_vocabulary for vocabulary.

**Comment by tomalec** [comment]
@PatrykSzwer  could you share "new designs"? The one you shared at https://github.com/Starcounter/uniform.css/issues/21#issue-295547192 has 20px

**Comment by alshakero** [comment]
>  they disregard entire rule 

Wow I didn't know that. I thought they would ignore that selector and continue. And apparently this is correct, at least for [FF (jsbin)](https://jsbin.com/ladayapine/1/edit?html,output).

**Comment by alshakero** [comment]
> Any opinions regarding #33 (comment)? Given the new single PR workflow, I'm not sure whether I should

I like this solution.

**Comment by tomalec** [comment]
> And apparently this is correct, at least for FF (jsbin).

FF stable (one that does not support `::slotted`) or Nightly (one that does)?

For me in FF stable it does not work.

**Comment by PatrykSzwer** [comment]
Guys, I've tried to separate things that I had to use more than couple of times, and most common was the solution with grid to provide correct width to section. Maybe we should somehow add this to uniform? As another version of section class, like `uni-primary-section-grid` or something like that (maybe something shorter)

File: src/uniform.css [hunk 8]
```diff
@@ -302,59 +302,43 @@ Pagination
...
 }
-.uni-alert > * {
-  padding: 15px;
-  margin-bottom: 20px;
-  border: 1px solid transparent;
-  border-radius: 4px;
+.uni-alert-info > .uni-alert-icons,
+.uni-alert-danger > .uni-alert-icons,
+.uni-alert-success > .uni-alert-icons,
+.uni-alert-warning > .uni-alert-icons {
```
Comment (warpech) [inline src/uniform.css]: I propose to change `.uni-alert-icons` to `.uni-alert-close` (as [Bootstrap](http://getbootstrap.com/docs/4.0/components/alerts/) does) or `.uni-alert-controls` (as [Windows Forms](https://superuser.com/questions/273335/is-there-a-name-for-the-close-minimize-and-restore-buttons) does).

`.uni-alert-icon` can be reserved to do a big icon on the left, such as: 

image

**Comment by warpech** [comment]
> @PatrykSzwer I had to use more than couple of times, and most common was the solution with grid to provide correct width to section. Maybe we should somehow add this to uniform? As another version of section class, like uni-primary-section-grid or something like that (maybe something shorter)

Moved to a separate issue: https://github.com/Starcounter/uniform.css/issues/37

----- 

@tomalec can you make a fix for https://github.com/Starcounter/uniform.css/pull/33#discussion_r168756719?

@alshakero can you make a commit to this PR that adds new CSS Custom Property names?

After that, this PR can be merged and we can release a new version.

File: src/uniform.css (hunk 8 above)
Comment (tomalec) [reply src/uniform.css]: I thought about close, but theoretically, it could be something like (i)-more info icon as well.

**Comment by tomalec** [comment]
Merging as EOF (End of Friday) is soon to come. 
@warpech  @alshakero  @PatrykSzwer  feel free to review retrospectively

**Comment by tomalec** [comment]
I had to rever `uni-alert` changes https://github.com/Starcounter/uniform.css/commit/b384a7ad12d733a8e90e34b6539bb0622110879d due to problems described at https://github.com/Starcounter/uniform.css/issues/21#issuecomment-366352985

**Pull request merged**